### PR TITLE
fix: Rework rollup config to output named exports (NODE-2712)

### PR DESCRIPTION
**Pull request by nbbeeken**
Rollup's output was bundling the library to have one single default export this change makes the output preserve the named outputs as well as have the default export. In addition the rollup config now compiles from typescript source instead of the typescript output.

NODE-2712

### CODE REVIEW

File: src/bson.ts [hunk 1]
```diff
@@ -277,3 +278,68 @@ export function deserializeStream(
...
+  serializeWithBufferAndIndex,
+  deserialize,
+  calculateObjectSize,
+  deserializeStream,
+  setInternalBufferSize,
+  // EJSON namespace
+  EJSON
+};
+
+export default bsonDefaultModuleExport;
```
Comment (mbroadst) [inline src/bson.ts:345]: Hmm, this looks pretty funky. Are we mixing named and default exports? It looks like you specified `named` in the rollup config, do you still need this here?

File: src/objectid.ts [hunk 2]
```diff
@@ -60,7 +60,7 @@ export class ObjectId {
       this.__id = id.__id;
     }
 
-    if (typeof id === 'object' && 'id' in id) {
+    if (typeof id === 'object' && id && 'id' in id) {
```
Comment (mbroadst) [inline src/objectid.ts:63]: why is this change needed? the `typeof` check should prove the existence of the object (it's not `typeof === undefined`)

File: rollup.config.js [hunk 3]
```diff
@@ -1,45 +1,72 @@
...
+        name: 'BSON',
+        sourcemap: true
       },
-      /* Browser IIFE bundle */
-      {
+      /* Browser IIFE bundle */ {
         file: 'dist/bson.bundle.js',
-        format: 'iife',
-        name: 'BSON'
+        format: 'esm',
```
Comment (mbroadst) [inline rollup.config.js:50]:
```suggestion
        format: 'iife',
```

File: rollup.config.js [hunk 4]
```diff
@@ -1,45 +1,72 @@
...
-      {
-        file: 'dist/bson.browser.esm.js',
-        format: 'es'
-      },
       /* Browser UMD bundle */
       {
         file: 'dist/bson.browser.umd.js',
-        format: 'umd',
-        name: 'BSON'
+        format: 'esm',
```
Comment (mbroadst) [inline rollup.config.js:44]:
```suggestion
        format: 'umd',
```

File: rollup.config.js [hunk 5]
```diff
@@ -1,45 +1,72 @@
...
+import typescript from '@rollup/plugin-typescript';
 
 // Entry point of the library
-const input = './lib/bson.js';
+const input = './src/bson.ts';
 
 export default [
+  {
+    input,
+    /* Browser ESM bundle */
```
Comment (mbroadst) [inline rollup.config.js:13]: can you explain what's going on in this change? Also, it looks like the config is mostly duplicated below

File: rollup.config.js [hunk 6]
```diff
@@ -49,13 +76,21 @@ export default [
...
+        sourcemap: true
+      }
+    ],
     // Notice the external buffer, and preferBuiltins false,
     // this bundle doesn't pull in 'buffer' or Map from 'core-js'
     external: ['buffer'],
-    plugins: [nodeResolve({ preferBuiltins: true }), commonjs()]
+    plugins: [
+      typescript({ tsconfig: './tsconfig.rollup.json' }),
+      nodeResolve({ preferBuiltins: true }),
```
Comment (mbroadst) [inline rollup.config.js:92]: I just noticed the comment above saying "note preferBuiltins false" which not actually the case here

File: tsconfig.rollup.json [hunk 7]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (mbroadst) [inline tsconfig.rollup.json:1]: does it make sense to embed this config directly in the `rollup.config.js`? I feel like otherwise it's just another file added to the heap of tooling files laying around. Is there a benefit to having this in its own file that I'm missing?

File: package.json [hunk 8]
```diff
@@ -32,6 +32,7 @@
     "@rollup/plugin-commonjs": "^15.0.0",
     "@rollup/plugin-json": "^4.1.0",
     "@rollup/plugin-node-resolve": "^9.0.0",
+    "@rollup/plugin-typescript": "^6.0.0",
```
Comment (mbroadst) [inline package.json:35]: was this change necessary for this fix?

**Review by mbroadst — CHANGES_REQUESTED** [review]

File: package.json (hunk 8 above)
Comment (nbbeeken) [reply package.json:35]: Yes, it fixed errors with the exports being incorrectly formatted since rollup was working off of the TS output it was translating from commonjs back to esm and this just made it easier to reason about the pipeline.

File: tsconfig.rollup.json (hunk 7 above)
Comment (nbbeeken) [reply tsconfig.rollup.json]: Yep sounds good!

File: rollup.config.js (hunk 6 above)
Comment (nbbeeken) [reply rollup.config.js]: Oops yea out of date comment, good catch.

File: rollup.config.js (hunk 5 above)
Comment (nbbeeken) [reply rollup.config.js:39]: Yea the key difference is the babel settings. There is definitely a lot of duplication but after wrestling with the config for a better part of this project an explicit set of settings for each output is actually quite helpful rather than debugging a programatic config.

File: rollup.config.js (hunk 4 above)
Comment (nbbeeken) [reply rollup.config.js:70]: Same as comment above. This is correct for what we're doing here.

File: rollup.config.js (hunk 3 above)
Comment (nbbeeken) [reply rollup.config.js:76]: The format here is correct, babel is downgrading the code after rollup has done its transformations so babel is handling the bundle format. I use UMD in the babel preset because it is equivalent to an iife (babel doesn't have an iife output without a plugin, which seems unnecessary).

File: src/objectid.ts (hunk 2 above)
Comment (nbbeeken) [reply src/objectid.ts:63]: `typeof null` is 'object', so this is checking for that

File: src/bson.ts (hunk 1 above)
Comment (nbbeeken) [reply src/bson.ts:345]: Yes, because our original bundles had both named and default exports so we need to keep the behavior.

**Review by reggi — APPROVED** [review]
LGTM, babel plugins to the rescue?

**Review by mbroadst — APPROVED** [review]
LGTM!

**Review by emadum — APPROVED** [review]
